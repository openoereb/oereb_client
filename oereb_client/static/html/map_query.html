<div ng-show="visible">
    <div class="loader-bg"></div>
    <div class="content" ng-show="contentVisible">
        <div class="list-group">
            <button type="button" class="list-group-item" ng-click="visible=false">
                <span class="pull-right"><strong>&times;</strong></span>
            </button>
            <button type="button"
                    class="list-group-item"
                    ng-repeat="re in realEstate"
                    ng-click="select(re.egrid)">
                Auszug f&uuml;r Parzelle {{re.number}}
            </button>
        </div>
    </div>
    <div class="loader" ng-show="!contentVisible">
        <i class="fa fa-spinner fa-pulse fa-2x"></i>
    </div>
    <div class="icon" ng-show="contentVisible">
        <div class="icon-outer">
            <div class="icon-inner"></div>
        </div>
    </div>
</div>
