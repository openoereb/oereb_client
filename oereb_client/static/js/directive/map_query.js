goog.provide('oereb.mapQueryDirective');

goog.require('oereb');
goog.require('oereb.MapService');
goog.require('oereb.EgridService');

/**
 * @function
 *
 * @description
 *
 * Directive definition function.
 *
 * @param {angular.$timeout} $timeout Angular $timeout service.
 * @param {oereb.MapService} MapService Service for map handling.
 * @param {oereb.EgridService} EgridService Service for querying EGRIDs.
 * @param {string} oerebEventEgridSelected Event name for selected EGRID.
 *
 * @returns {angular.Directive} Angular directive definition.
 *
 * @ngInject
 */
oereb.mapQueryDirective = function($timeout, MapService, EgridService, oerebEventEgridSelected) {
  return {
    restrict: 'E',
    replace: true,
    templateUrl: 'static/html/map_query.html',
    scope: {},
    link: function(scope, element) {
      var centerRealEstate = false;

      /** @export {boolean} */
      scope.visible = false;

      /** @export {boolean} */
      scope.contentVisible = false;

      /** @export {Array} */
      scope.realEstate = [];

      // Get map instance
      scope.map_ = MapService.getMap();

      // Create overlay for popup
      scope.overlay_ = new ol.Overlay({
        position: [0, 0],
        element: element[0]
      });

      scope.map_.addOverlay(scope.overlay_);

      // Add click listener
      scope.map_.on('singleclick', function(evt) {
        var coord = scope.map_.getEventCoordinate(evt.originalEvent);
        scope.queryAt(coord);
      });

      /**
       * @ngdoc method
       * @name oerebMapQuery#queryAt
       *
       * @description
       *
       * Routine to provide the item picker for selecting unique egrid.
       *
       * @param {ol.Coordinate} coord The position to move to element to.
       * @param {boolean} center Switch for centering map view on the selector or not.
       *
       * @export
       */
      scope.queryAt = function(coord, center) {
        scope.moveTo_(coord);
        scope.queryEgrid_(coord);
        centerRealEstate = center;
        if (center) {
          scope.map_.getView().setCenter(coord);
        }
      };

      /**
       * @ngdoc method
       * @name oerebMapQuery#moveTo_
       *
       * @description
       *
       * Move map query to pixel position and show loading indicator.
       *
       * @param {ol.Coordinate} coord The position to move to element to.
       *
       * @private
       */
      scope.moveTo_ = function(coord) {
        $timeout(function() {
          scope.visible = true;
          scope.contentVisible = false;
          scope.overlay_.setPosition(coord);
        });
      };

      /**
       * @ngdoc method
       * @name oerebMapQuery#queryEgrid_
       *
       * @description
       *
       * Query EGRID for the specified coordinate.
       *
       * @param {ol.Coordinate} coord The coordinate to query the EGRID for.
       *
       * @private
       */
      scope.queryEgrid_ = function(coord) {
        EgridService.getEgridByCoord(coord).then(
          function(data) {
            scope.realEstate = data;
            if (scope.realEstate.length < 2) {
              if (scope.realEstate.length === 1) {
                scope.$emit(oerebEventEgridSelected, scope.realEstate[0]['egrid'], centerRealEstate);
              }
              $timeout(function () {
                scope.contentVisible = false;
                scope.visible = false;
              });
            }
            else {
              $timeout(function () {
                scope.contentVisible = true;
              });
            }
          },
          function() {
            scope.realEstate = [];
            $timeout(function() {
              scope.contentVisible = false;
              scope.visible = false;
            });
            // TODO: display error message
          }
        );
      };

      /**
       * @ngdoc method
       * @name oerebMapQuery#close
       *
       * @description
       *
       * Close the map query.
       *
       * @export
       */
      scope.close = function() {
        $timeout(function() {
          scope.contentVisible = false;
          scope.visible = false;
        });
      };

      /**
       * @ngdoc method
       * @name oerebMapQuery#select
       *
       * @description
       *
       * Load the extract for the specified EGRID. It uses the switch to recenter the map from
       * centerRealEstate variable.
       *
       * @param {string} egrid The EGRID to load the extract for.
       *
       * @export
       */
      scope.select = function(egrid) {
        $timeout(function() {
          scope.visible = false;
        });
        scope.$emit(oerebEventEgridSelected, egrid, centerRealEstate);
      };

    }
  }
};

/**
 * @ngdoc directive
 * @name oerebMapQuery
 * @module oereb
 * @restrict E
 *
 * @description
 *
 * Query and select EGRIDs on the map.
 */
oereb.module.directive('oerebMapQuery', oereb.mapQueryDirective);
