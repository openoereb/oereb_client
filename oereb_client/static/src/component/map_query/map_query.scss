@import "../../oereb_client.scss";

$map-query-shadow: $default-shadow;
$map-query-loader-size: 34px;
$map-query-loader-bg: #fff;
$map-query-content-bg: $map-query-loader-bg;
$map-query-content-width: 250px;
$map-query-icon-color: #f00;
$map-query-icon-line-width: 4px;

.oereb-client-overlay {
  position: absolute;
  overflow: visible;

  div.loader {
    width: $map-query-loader-size;
    height: $map-query-loader-size;
    border-radius: 50%;
    background-color: $map-query-loader-bg;
    box-shadow: $map-query-shadow;

    .spinner-grow {
      width: $map-query-loader-size !important;
      height: $map-query-loader-size !important;
    }
  }

  div.results {
    div.background-icon {
      position: absolute;
      z-index: 1;
      width: $map-query-loader-size;
      height: $map-query-loader-size;
      border-radius: 50%;
      background-color: $map-query-loader-bg;
      box-shadow: $map-query-shadow;
    }

    div.content {
      position: absolute;
      z-index: 2;
      margin-top: $map-query-loader-size/2;
      margin-left: $map-query-loader-size/2;
      background-color: $map-query-loader-bg;
      box-shadow: $map-query-shadow;
      width: $map-query-content-width;

      button.list-group-item {
        padding-left: $map-query-loader-size/2 + 15px;
        padding-right: 10px;
        padding-top: 3px;
        padding-bottom: 3px;
      }
    }

    div.icon {
      position: absolute;
      z-index: 3;
      width: $map-query-loader-size;
      height: $map-query-loader-size;
      border-radius: 50%;
      background-color: transparent;

      div.icon-outer {
        position: relative;
        width: $map-query-loader-size;
        height: $map-query-loader-size;
        border-style: solid;
        border-width: $map-query-icon-line-width;
        border-color: $map-query-icon-color;
        border-radius: $map-query-loader-size/2;
        background-color: $map-query-loader-bg;

        div.icon-inner {
          position: relative;
          background-color: $map-query-icon-color;
          border-radius: ($map-query-loader-size - 4*$map-query-icon-line-width) / 2;
          width: $map-query-loader-size - 4*$map-query-icon-line-width;
          height: $map-query-loader-size - 4*$map-query-icon-line-width;
          top: $map-query-icon-line-width;
          left: $map-query-icon-line-width;
        }
      }
    }
  }
}
