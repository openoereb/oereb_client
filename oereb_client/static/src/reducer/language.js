import {createSlice} from "@reduxjs/toolkit";

import i18n from "../i18n";

export const languageSlice = createSlice({
  name: 'language',
  initialState: {
    current: null,
    default: null,
    available: []
  },
  reducers: {
    initLanguages: (state, action) => {
      state.default = action.payload.default;
      state.available = action.payload.available;
      const query = new URLSearchParams(window.location.search);
      if (query.has('lang') && state.available.indexOf(query.get('lang')) > -1) {
        state.current = query.get('lang');
      }
      else {
        state.current = state.default;
      }
      i18n.changeLanguage(state.current);
      query.set('lang', state.current);
      window.history.pushState(null, null, '?' + query.toString());
    },
    setLanguage: (state, action) => {
      if (state.available.indexOf(action.payload) > -1) {
        state.current = action.payload;
        i18n.changeLanguage(state.current);
        const query = new URLSearchParams(window.location.search);
        query.set('lang', state.current);
        window.history.pushState(null, null, '?' + query.toString());
      }
    }
  }
});

export const {initLanguages, setLanguage} = languageSlice.actions;

export default languageSlice.reducer;
