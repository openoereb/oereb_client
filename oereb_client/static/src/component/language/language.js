import PropTypes from 'prop-types';
import React from "react";
import {withNamespaces} from 'react-i18next';
import {useDispatch, useSelector} from "react-redux";

import {setLanguage} from "../../reducer/language";

const OerebLanguage = function({t}) {
  const dispatch = useDispatch();
  const language = useSelector((state) => state.language);
  const currentLanguage = language.current;
  const availableLanguages = language.available;

  const selectLanguage = function(lang) {
    dispatch(setLanguage(lang));
  };

  const languages = availableLanguages.map((language, key) =>
    <li key={key}>
      <button className="dropdown-item" onClick={selectLanguage.bind(this, language)}>
        {language.toUpperCase()}
      </button>
    </li>
  );

  return (
    <div>
      <button className="btn btn-outline-secondary dropdown-toggle"
        type="button"
        data-bs-toggle="dropdown"
        aria-expanded="false"
        title={t('menu.language.title')}>
        {currentLanguage.toUpperCase()}
      </button>
      <ul className="dropdown-menu">
        {languages}
      </ul>
    </div>
  );
};

OerebLanguage.propTypes = {
  t: PropTypes.func.isRequired
};

export default withNamespaces()(OerebLanguage);
