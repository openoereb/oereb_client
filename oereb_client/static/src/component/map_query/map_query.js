import './map_query.scss';

import Overlay from 'ol/Overlay';
import PropTypes from 'prop-types';
import React, {useRef} from 'react';
import {useDispatch, useSelector} from 'react-redux';

import {queryExtractById} from '../../api/extract';
import {setViewServices} from '../../reducer/accordion';
import {loadExtract, showError, showExtract} from '../../reducer/extract';
import {updateHistory} from '../../reducer/history';
import {hide} from '../../reducer/map_query';

const OerebMapQuery = function(props) {
    const dispatch = useDispatch();
    const config = useSelector((state) => state.config).config;
    const mapQuery = useSelector((state) => state.mapQuery);
    const mapQueryElement = useRef(null);

    const applicationUrl = config.application_url;
    const map = props.map;

    const overlay = new Overlay({
        element: mapQueryElement.current,
        autoPan: true,
        offset: [-17, -17]
    });
    map.addOverlay(overlay);
    overlay.setPosition([mapQuery.posX, mapQuery.posY]);

    const close = function() {
        dispatch(hide());
    };

    const queryExtract = function(egrid) {
        dispatch(hide());
        dispatch(setViewServices([]));
        dispatch(loadExtract(egrid));
        queryExtractById(applicationUrl, egrid)
        .then((extract) => {
            dispatch(showExtract(extract));
            dispatch(updateHistory(extract));
        })
        .catch(() => {
            dispatch(showError());
        });
    };

    const listResults = mapQuery.results.map((result, key) => {
        const number = result.number;
        const egrid = result.egrid;
        return (
            <button key={key}
                    onClick={queryExtract.bind(this, egrid)}
                    type="button"
                    className="list-group-item list-group-item-action">
                <small>Auszug für Grundstück {number}</small>
            </button>
        );
    });

    if (mapQuery.loading) {
        return (
            <div className="oereb-client-overlay" ref={mapQueryElement}>
                <div className="loader">
                    <div className="spinner-grow"></div>
                </div>
            </div>
        );
    }
    else if (mapQuery.visible) {
        return (
            <div className="oereb-client-overlay" ref={mapQueryElement}>
                <div className="results">
                    <div className="background-icon"></div>
                    <div className="content list-group">
                        <button onClick={close} type="button" className="list-group-item list-group-item-action text-end">
                            <strong className="bi bi-x"></strong>
                        </button>
                        {listResults}
                    </div>
                    <div className="icon">
                        <div className="icon-outer">
                            <div className="icon-inner"></div>
                        </div>
                    </div>
                </div>
            </div>
        );
    }

        return (
            <div className="oereb-client-overlay" ref={mapQueryElement}></div>
        );

};

OerebMapQuery.propTypes = {
    map: PropTypes.object.isRequired
};

export default OerebMapQuery;
