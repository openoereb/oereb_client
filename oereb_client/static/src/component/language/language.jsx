import React from "react";
import {useTranslation} from "react-i18next";
import {useDispatch, useSelector} from "react-redux";

import {hideExtract, loadExtract, showError, showExtract} from "../../reducer/extract";
import {updateHistory} from "../../reducer/history";
import {setLanguage} from "../../reducer/language";
import {queryExtractById} from "../../request/extract";
import {TooManyRequestsError} from "../../util/error";
import {showWarning} from "../../reducer/message";

/**
 * This component provides the language selector.
 */
const OerebLanguage = function() {
  const {t} = useTranslation();
  const dispatch = useDispatch();
  const config = useSelector((state) => state.config).config;
  const serviceUrl = config.service_url;
  const language = useSelector((state) => state.language);
  const currentLanguage = language.current;
  const availableLanguages = language.available;
  const extract = useSelector((state) => state.extract);

  const selectLanguage = function(lang) {
    if (lang !== currentLanguage) {
      dispatch(setLanguage(lang));
      if (extract.visible) {
        dispatch(loadExtract({
          egrid: extract.egrid,
          identdn: extract.identdn,
          number: extract.number,
          zoom: false
        }));
        queryExtractById(
          serviceUrl,
          extract.egrid,
          extract.identdn,
          extract.number,
          config.extract_json_timeout,
          lang
        )
          .then((data) => {
            dispatch(showExtract(data));
            dispatch(updateHistory(data));
          })
          .catch((error) => {
            if (error instanceof TooManyRequestsError) {
              dispatch(hideExtract());
              dispatch(showWarning(t('extract.error.too_many_requests'), true));
            }
            else {
              dispatch(showError());
            }
          });
      }
    }
  };

  const languages = availableLanguages.map((language, key) =>
    <li key={key}>
      <button className="dropdown-item" onClick={selectLanguage.bind(this, language)}>
        {language.toUpperCase()}
      </button>
    </li>
  );

  return (
    <div className="oereb-client-language-selector">
      <button className="btn btn-outline-secondary dropdown-toggle"
        type="button"
        data-bs-toggle="dropdown"
        aria-expanded="false"
        title={t("menu.language.title")}>
        {currentLanguage.toUpperCase()}
      </button>
      <ul className="dropdown-menu">
        {languages}
      </ul>
    </div>
  );
};

export default OerebLanguage;
