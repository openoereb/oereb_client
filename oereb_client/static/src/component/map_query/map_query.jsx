import './map_query.scss';

import Overlay from 'ol/Overlay';
import PropTypes from 'prop-types';
import React, {useEffect, useRef, useState} from 'react';
import {useDispatch, useSelector} from 'react-redux';

import {setViewServices} from '../../reducer/accordion';
import {loadExtract, showError, showExtract} from '../../reducer/extract';
import {updateHistory} from '../../reducer/history';
import {hide} from '../../reducer/map_query';
import {queryExtractById} from '../../request/extract';
import {getLocalizedText} from '../../util/language';

/**
 * A OpenLayers overlay component to select a real estate in the map.
 */
const OerebMapQuery = function (props) {
  const dispatch = useDispatch();
  const config = useSelector((state) => state.config).config;
  const mapQuery = useSelector((state) => state.mapQuery);
  const mapQueryElement = useRef(null);
  const language = useSelector((state) => state.language);
  const currentLanguage = language.current;
  const defaultLanguage = language.default;
  const [overlay, setOverlay] = useState(null);

  const serviceUrl = config.service_url;
  const map = props.map;

  useEffect(() => {
    if (overlay === null) {
      const newOverlay = new Overlay({
        element: mapQueryElement.current,
        autoPan: {
          animation: true,
          margin: Math.min(document.body.offsetWidth, document.body.offsetHeight) / 2
        },
        offset: [-17, -17]
      });
      map.addOverlay(newOverlay);
      setOverlay(newOverlay);
    }
  });

  const close = function () {
    dispatch(hide());
  };

  const queryExtract = function (egrid, identdn, number) {
    dispatch(hide());
    dispatch(setViewServices([]));
    dispatch(loadExtract({
      egrid: egrid,
      identdn: identdn,
      number: number,
      zoom: false
    }));
    queryExtractById(serviceUrl, egrid, identdn, number, config.extract_json_timeout, currentLanguage)
      .then((extract) => {
        dispatch(showExtract(extract));
        dispatch(updateHistory(extract));
      })
      .catch(() => {
        dispatch(showError());
      });
  };

  const listResults = mapQuery.results.map((result, key) => {
    const egrid = result.egrid;
    const identdn = result.identDN;
    const number = result.number;
    const type = getLocalizedText(result.type.Text, currentLanguage, defaultLanguage);
    return (
      <button key={key}
        onClick={queryExtract.bind(this, egrid, identdn, number)}
        type="button"
        className="list-group-item list-group-item-action">
        <small>{type} {number}</small>
      </button>
    );
  });

  if (mapQuery.loading) {
    overlay.setPosition([mapQuery.posX, mapQuery.posY]);
    return (
      <div className="oereb-client-overlay" ref={mapQueryElement}>
        <div className="loader">
          <div className="spinner-grow"></div>
        </div>
      </div>
    );
  }
  else if (mapQuery.visible) {
    overlay.setPosition([mapQuery.posX, mapQuery.posY]);
    return (
      <div className="oereb-client-overlay" ref={mapQueryElement}>
        <div className="results">
          <div className="background-icon"></div>
          <div className="content list-group">
            <button onClick={close}
              type="button"
              className="list-group-item list-group-item-action text-end">
              <strong className="bi bi-x"></strong>
            </button>
            {listResults}
          </div>
          <div className="icon">
            <div className="icon-outer">
              <div className="icon-inner"></div>
            </div>
          </div>
        </div>
      </div>
    );
  }

  return (
    <div className="oereb-client-overlay" ref={mapQueryElement}></div>
  );

};

OerebMapQuery.propTypes = {

  /** The OpenLayers map instance. */
  map: PropTypes.object.isRequired

};

export default OerebMapQuery;
