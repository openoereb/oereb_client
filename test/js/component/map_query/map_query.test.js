import {render} from '@testing-library/react';
import {Map} from "ol";
import React from "react";
import {act} from "react-dom/test-utils";
import {Provider} from "react-redux";

import OerebMapQuery from "../../../../oereb_client/static/src/component/map_query/map_query";
import {initLanguages} from "../../../../oereb_client/static/src/reducer/language";
import {loadAt, show} from "../../../../oereb_client/static/src/reducer/map_query";
import MainStore from "../../../../oereb_client/static/src/store/main";

describe('map query component', () => {

  const map = new Map();

  let component;

  beforeEach(() => {
    act(() => {
      MainStore.dispatch(initLanguages({
        default: 'de',
        available: ['de']
      }));
    });
    component = render(
      <Provider store={MainStore}>
        <OerebMapQuery map={map} />
      </Provider>
    );
  });

  it('should render map query', () => {
    expect(component.asFragment()).toMatchSnapshot();
    act(() => {
      MainStore.dispatch(loadAt({
        posX: 0.0,
        posY: 0.0
      }));
    });
    expect(component.asFragment()).toMatchSnapshot();
    act(() => {
      MainStore.dispatch(show({
        results: [
          {
            egrid: 'CH0815',
            number: '0815',
            identDN: 'SAMPLE1',
            type: {
              Code: 'Distinct_and_permanent_rights.BuildingRight',
              Text: [
                {
                  Language: 'de',
                  Text: 'Baurecht'
                },
                {
                  Language: 'fr',
                  Text: 'Droit de superficie'
                },
                {
                  Language: 'it',
                  Text: 'Diritto di superficie'
                },
                {
                  Language: 'rm',
                  Text: 'Dretg da construcziun'
                },
                {
                  Language: 'en',
                  Text: 'Building right'
                }
              ]
            }
          },
          {
            egrid: 'CH1234',
            number: '1234',
            identDN: 'SAMPLE2',
            type: {
              Code: 'RealEstate',
              Text: [
                {
                  Language: 'de',
                  Text: 'Liegenschaft'
                },
                {
                  Language: 'fr',
                  Text: 'Bien-fonds'
                },
                {
                  Language: 'it',
                  Text: 'Bene immobile'
                },
                {
                  Language: 'rm',
                  Text: 'Bain immobigliar'
                },
                {
                  Language: 'en',
                  Text: 'Property'
                }
              ]
            }
          }
        ]
      }));
    });
    expect(component.asFragment()).toMatchSnapshot();
  });

});