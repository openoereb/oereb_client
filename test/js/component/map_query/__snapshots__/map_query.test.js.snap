// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`map query component should render map query 1`] = `<DocumentFragment />`;

exports[`map query component should render map query 2`] = `<DocumentFragment />`;

exports[`map query component should render map query 3`] = `<DocumentFragment />`;
